### PR TITLE
HPCC-22008 Fix race condition in non-linking cache implementations 

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
Pushed to run smoketest, but believed correct
I'm not sure whether to target 7.0 or 7.2 - targeting 7.2 for the moment
Final release should remove the critical section that Jake added, and can also remove setting xxcount in the destructor since that will no longer highlight the issue
@mckellyln @jakesmith @richardkchapman please review.

File: roxie/ccd/ccdfile.cpp [hunk 1]
```diff
@@ -998,9 +998,9 @@ class CRoxieFileCache : implements IRoxieFileCache, implements ICopyFileProgress
                     if (todo.ordinality())
                     {
                         ILazyFileIO *popped = &todo.popGet();
-                        if (popped->isAlive())
+                        if (popped->isAliveAndLink())
```
Comment (ghalliday) [inline roxie/ccd/ccdfile.cpp]: I think this could have caused problems - I vaguely remember crashes in this area.

**Comment by ghalliday** [comment]
@mckellyln can you take a look at this?

File: ecl/hql/hqlgram2.cpp [hunk 2]
```diff
@@ -12686,13 +12686,14 @@ bool verifySimplifiedDefinition(IHqlExpression *origExpr, IHqlExpression *simpli
...
-    printf("Sizes: const(%u) expr(%u) select(%u) dataset(%u) annotation(%u) prop(%u)\n",
+    printf("Sizes: const(%u) expr(%u) select(%u) dataset(%u) annotation(%u) prop(%u) loc(%u)\n",
             (unsigned)sizeof(CHqlConstant),
             (unsigned)sizeof(CHqlExpressionWithType),
             (unsigned)sizeof(CHqlSelectExpression),
             (unsigned)sizeof(CHqlDataset),
             (unsigned)sizeof(CHqlAnnotation),
-            (unsigned)sizeof(CHqlDynamicProperty)
+            (unsigned)sizeof(CHqlDynamicProperty),
+            (unsigned)sizeof(CHqlLocationAnnotation)
```
Comment (mckellyln) [inline ecl/hql/hqlgram2.cpp]: Is this file change really part of the same issue ?

File: roxie/ccd/ccdfile.cpp [hunk 3]
```diff
@@ -1279,24 +1280,28 @@ class CRoxieFileCache : implements IRoxieFileCache, implements ICopyFileProgress
...
+                            // That's OK here, since we know the file is open and remote.
+                            // But don't be tempted to move this line outside these if's (eg. to trace the idle case)
+                            const char *fname = remote ? f->querySource()->queryFilename() : f->queryFilename();
+                            DBGLOG("Closing inactive %s file %s (last accessed %u ms ago)", remote ? "remote" : "local",  fname, age);
+                        }
+                        f->close();
                     }
-                    f->close();
+                    else
+                        goers.append(*f);   //Shouldn't this be linked?
```
Comment (mckellyln) [inline roxie/ccd/ccdfile.cpp]: It is linked now once.  Do you mean add a LINK(*f) ?

File: system/jlib/jiface.hpp [hunk 4]
```diff
@@ -68,7 +68,7 @@ class CSimpleInterfaceOf : public INTERFACE
 {
     friend class CInterfaceOf<INTERFACE>;    // want to keep xxcount private outside this pair of classes
 public:
-    inline virtual ~CSimpleInterfaceOf() {}
+    inline virtual ~CSimpleInterfaceOf() { xxcount.store(0, std::memory_order_relaxed); }
```
Comment (mckellyln) [inline system/jlib/jiface.hpp]: Should this set xxcount to 0 or decrement it by 1 ?

File: system/jlib/jiface.hpp [hunk 5]
```diff
@@ -119,31 +119,46 @@ class CInterfaceOf : public CSimpleInterfaceOf<INTERFACE>
...
+    //must be called inside a critical section that protects access to the cache.
+    //This function must only increment the link count if the object is valid - otherwise it
+    //end up being freed more than once.
+    inline bool isAliveAndLink() const
+    {
+        unsigned expected = this->xxcount.load(std::memory_order_acquire);
+        for (;;)
+        {
+            //If count is 0, or count >= DEAD_PSEUDO_COUNT then return false - combine both into a single test
+            if ((expected-1) >= (DEAD_PSEUDO_COUNT-1))
```
Comment (mckellyln) [inline system/jlib/jiface.hpp]: clever

File: system/jlib/jiface.hpp (hunk 4 above)
Comment (ghalliday) [reply system/jlib/jiface.hpp]: This can be removed - it was used to make the issue more reproducible, but has no benefit with the new scheme.

File: system/jlib/jiface.hpp [hunk 6]
```diff
@@ -119,31 +119,46 @@ class CInterfaceOf : public CSimpleInterfaceOf<INTERFACE>
...
     {
         if (this->xxcount.fetch_sub(1,std::memory_order_release) == 1)
         {
-            unsigned zero = 0;
-            //Because beforeDispose could cause this object to be linked/released or call isAlive(), this->xxcount is set
-            //to a a high mid-point positive number to avoid poss. of releasing again.
-            if (this->xxcount.compare_exchange_strong(zero, DEAD_PSEUDO_COUNT, std::memory_order_acq_rel))
+            //Overwrite with a special value so that any calls to Link()/Release within beforeDispose()
+            //do not cause the object to be released again.
+            this->xxcount.store(DEAD_PSEUDO_COUNT, std::memory_order_release);
```
Comment (mckellyln) [inline system/jlib/jiface.hpp]: I think I like the old method here - if count is zero then atomically exchange it with DEAD_PSEUDO_COUNT and call beforeDispose().
What prevents someone else from linking between the fetch_sub() and store() ?

**Comment by mckellyln** [comment]
@ghalliday looks good - but do have one question above.

File: system/jlib/jiface.hpp (hunk 6 above)
Comment (mckellyln) [reply system/jlib/jiface.hpp]: I think if LINK() is never called on it, but only isAliveAndLink() then its ok.

Comment (ghalliday) [reply system/jlib/jiface.hpp]: If link is called, then you are back to the problem of having a potential race condition.

File: roxie/ccd/ccdfile.cpp [hunk 7]
```diff
@@ -1275,28 +1276,32 @@ class CRoxieFileCache : implements IRoxieFileCache, implements ICopyFileProgress
...
                     }
-                    f->close();
+                    else
+                        goers.append(*f.getClear());
                 }
-                else
-                    goers.append(*f);
             }
         }
         unsigned numFilesLeft = goers.ordinality();
```
Comment (ghalliday) [inline roxie/ccd/ccdfile.cpp]: @richardkchapman if goers are now linked could the critical section finish here instead?

File: system/jlib/jiface.hpp (hunk 6 above)
Comment (ghalliday) [reply system/jlib/jiface.hpp]: and cas is much more expensive than a store.

**Comment by ghalliday** [comment]
Also updated incorrect comment in jiface

**Comment by ghalliday** [comment]
@richardkchapman please review very carefully

File: system/jlib/jiface.hpp [hunk 8]
```diff
@@ -119,31 +119,46 @@ class CInterfaceOf : public CSimpleInterfaceOf<INTERFACE>
...
+    inline bool isAliveAndLink() const
+    {
+        unsigned expected = this->xxcount.load(std::memory_order_acquire);
+        for (;;)
+        {
+            //If count is 0, or count >= DEAD_PSEUDO_COUNT then return false - combine both into a single test
+            if ((expected-1) >= (DEAD_PSEUDO_COUNT-1))
+                return false;
+
+            //Avoid incrementing the link count if expected=0, otherwise it introduces a race condition
```
Comment (jakesmith) [inline system/jlib/jiface.hpp]: the comment confused me a little, as expected=0 already dealt with above, perhaps comment needs changing to just:
```
  // compare and exchange to increment to avoid race condition
```

.. or similar

File: system/jlib/jiface.hpp (hunk 4 above)
Comment (jakesmith) [reply system/jlib/jiface.hpp]: just commenting as reminder still needs removing.

File: system/jlib/jiface.hpp [hunk 9]
```diff
@@ -219,7 +234,11 @@ class CSingleThreadInterfaceOf : public CSingleThreadSimpleInterfaceOf<INTERFACE
 public:
     virtual void beforeDispose() {}
 
-    inline bool isAlive() const         { return this->xxcount < DEAD_PSEUDO_COUNT; }       //only safe if Link() is called first
+    inline bool isAliveAndLink() const
+    {
+        this->Link();
+        return this->xxcount < DEAD_PSEUDO_COUNT;
+    }
```
Comment (jakesmith) [inline system/jlib/jiface.hpp]: does this method (and isAlive() before it), make sense given it's suppose to be SingleThreaded? Is it used?

File: fs/dafsclient/rmtfile.cpp [hunk 10]
```diff
@@ -616,13 +616,13 @@ class CCritTable : private SimpleHashTableOf<CEndpointCS, const SocketEndpoint>
     CEndpointCS *getCrit(const SocketEndpoint &ep)
     {
         CriticalBlock b(crit);
-        Linked<CEndpointCS> clientCrit = find(ep);
-        if (!clientCrit || !clientCrit->isAlive()) // if !isAlive(), then it is in the process of being destroyed/removed.
+        CEndpointCS * clientCrit = find(ep);
+        if (!clientCrit || !clientCrit->isAliveAndLink()) // if !isAlive(), then it is in the process of being destroyed/removed.
```
Comment (jakesmith) [inline fs/dafsclient/rmtfile.cpp]: commend should be updated: // if !isAliveAndLink(),

**Comment by jakesmith** [comment]
@ghalliday - looks good, a couple of trivial comments/questions.

File: system/jlib/jiface.hpp (hunk 8 above)
Comment (ghalliday) [reply system/jlib/jiface.hpp]: yes, should say if xxcount=0.  because the value may have changed to 0 in the meantime.
I will modify

File: system/jlib/jiface.hpp (hunk 9 above)
Comment (ghalliday) [reply system/jlib/jiface.hpp]: I think it is here so that the class can be used in the code generator without modifying any other code.

File: fs/dafsclient/rmtfile.cpp (hunk 10 above)
Comment (ghalliday) [reply fs/dafsclient/rmtfile.cpp]: done

File: system/jlib/jiface.hpp (hunk 8 above)
Comment (ghalliday) [reply system/jlib/jiface.hpp]: changed comment

File: system/jlib/jiface.hpp (hunk 4 above)
Comment (ghalliday) [reply system/jlib/jiface.hpp]: now removed

File: roxie/ccd/ccdfile.cpp (hunk 3 above)
Comment (ghalliday) [reply roxie/ccd/ccdfile.cpp]: fixed

**Comment by ghalliday** [comment]
@richardkchapman please review.  Also one question for you..

File: ecl/hql/hqlgram2.cpp (hunk 2 above)
Comment (richardkchapman) [inline ecl/hql/hqlgram2.cpp]: This seems unrelated (though harmless) ?

File: roxie/ccd/ccdfile.cpp (hunk 7 above)
Comment (richardkchapman) [reply roxie/ccd/ccdfile.cpp]: I would have thought so

File: system/jlib/jiface.hpp [hunk 11]
```diff
@@ -119,31 +119,46 @@ class CInterfaceOf : public CSimpleInterfaceOf<INTERFACE>
 public:
     virtual void beforeDispose() {}
 
-    inline bool isAlive() const         { return this->xxcount.load(std::memory_order_relaxed) < DEAD_PSEUDO_COUNT; }       //only safe if Link() is called first
+    //Function to be called when checking if an entry in a non-linking cache is valid
+    //must be called inside a critical section that protects access to the cache.
+    //This function must only increment the link count if the object is valid - otherwise it
+    //end up being freed more than once.
```
Comment (richardkchapman) [inline system/jlib/jiface.hpp]: trivial: grammar (end should say ends)

File: system/jlib/jiface.hpp (hunk 5 above)
Comment (richardkchapman) [reply system/jlib/jiface.hpp]: Too clever? Is this going to fall foul of new compilers being more aggressive regarding overflow/underflow optimizations, or is it ok because unsigned overflow/underflow is well-defined?

**Review by richardkchapman — COMMENTED** [review]
@ghalliday Some trivial comments, did not spot anything major

File: ecl/hql/hqlgram2.cpp (hunk 2 above)
Comment (ghalliday) [reply ecl/hql/hqlgram2.cpp]: I can't remember why I added this - I wanted to check what its size was, but can't remember why.

File: system/jlib/jiface.hpp (hunk 5 above)
Comment (ghalliday) [reply system/jlib/jiface.hpp]: unsigned arithmetic is well defined (modulo 2^n) so will not have any problems.

**Review by ghalliday — COMMENTED** [review]
@richardkchapman

File: system/jlib/jiface.hpp (hunk 11 above)
Comment (ghalliday) [reply system/jlib/jiface.hpp]: fixed
